### PR TITLE
RSE Blog post: xarray for earth

### PR DESCRIPTION
**Pull request by jmcook1186**
Hi,

This pull request is for a post about xarray for the RSE blog, as discussed with Will F.
Please review and let me know any amends.

3 files committed: 
2019-02-08-xarray-for-earth.md to /posts
Sentinel_map.jpg to /assets/images/
Sentinel_map2.jpg to assets/images

### CODE REVIEW

File: _posts/2019-02-08-xarray-for-earth.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+Loading a NetCDF file is as simple as `x = xr.open_dataset(netcdffile)` and saving a dataframe or array to NetCDF is equally straighforward: `DF.to_netCDF('filename')`.
+
+### Well suited to parallel computing using Dask
+Xarray is also tightly integrated with Dask so that operations over datasets too large to sit in memory can be achieved using Dask commands directly over xarray data arrays or datasets. Dask achieves this by chunking the dataset and distributing the computation over multiple cores. xarray takes a “chunks” argument that divides the xarray into dask arrays ready to be distributed. More information about xarray and dask is avilable in the (very good) [documentation](http://xarray.pydata.org/en/stable/dask.html) .  
+
+
+## Data structures:
+
+xarray works with two basic data structures – the DataArray and the DataSet
```
Comment (ptheywood) [inline _posts/2019-02-08-xarray-for-earth.md]: Missing `.`

File: _posts/2019-02-08-xarray-for-earth.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+### DataSet
+A DataSet is a collection of DataArrays. The power of the Dataset structure comes from the ability to perform operations over multiple arrays along shared dimensions.
+
+## Example
+In this example I use xarray to load in a dataset from a NetCDF file. The dataset includes layers representing 2D maps of a section of the Greenland Ice Sheet imaged using the ESA satellite Sentinel-2. Sentinel-2 images the surface in nine wavelengths ('bands) at 20m ground resolution so there are actually nine layers for each image, each one projected identically in space (there was some preprocessing of these data using GDAL that I will not cover here). The layers are loaded in using xarray, converted to numpy arrays and processed, then saved to a new NetCDF with metadata and attributes. xarray’s plot function is used to display the key layers.
+
+First the necessary packages are imported and the image bands are loaded in from a NetCDF file using xarray. Only the reflectance values are extracted from the data array and saved to a 9D numpy array, where each dimension represents a wavelength.
+
+```
```
Comment (ptheywood) [inline _posts/2019-02-08-xarray-for-earth.md]: Syntax highlghting can be applied via:

    ```python

File: _posts/2019-02-08-xarray-for-earth.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+for i in np.arange(0,len(bands),1): #loop through bands in NetCDF file
+    S2BX = xr.open_dataset(str(img_path+'B'+bands[i]+'.nc')) # open only the relevant band from the NetCDF
+    S2BXarr = S2BX.to_array() # send the band to a numpy array 
+    S2BXvals = np.array(S2BXarr.variable.values[1]) # extract only the reflectance values
+    S2BXvals = S2BXvals.astype(float) #convert to float
+    S2vals[i,:,:] = S2BXvals # append to 'master' ND array
+vals = vals/10000 # correct unit from S2 L2A data to reflectance between 0-1
+```
+In the next stage, a scikit-learn classifier is applied to the 9D array. This classifier has been trained in a separate script, saved to a .pkl file and then loaded in here. The 9D array is also converted into a layer that applies a conversion formula to predict the albedo -or reflectivity - of the surface. This is a stripped down version of a code developed at www.github.com/jmcook1186/IceSurfClassifiers/ - go there for the complete scripts.
+```
```
Comment (ptheywood) [inline _posts/2019-02-08-xarray-for-earth.md]: Syntax highlghting can be applied via:

    ```python

File: _posts/2019-02-08-xarray-for-earth.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    predicted[predicted == 'HA'] = float(6)
+    predicted[predicted == 'Zero'] = float(0)
+
+    # reshape 1D array back into original image dimensions
+    predicted = np.reshape(predicted, [lenx, leny])
+    albedo = np.reshape(albedo_array, [lenx, leny])
+```    
+There is now a classified surface map and an albedo map in memory. The next section shows how this is collated into an xarray dataset with some metadata and saved to a new NetCDF file.
+   
+```
```
Comment (ptheywood) [inline _posts/2019-02-08-xarray-for-earth.md]: Syntax highlghting can be applied via:

    ```python

File: _posts/2019-02-08-xarray-for-earth.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    dataset.to_netcdf(savefig_path + "Classification_and_Albedo_Data_{}.nc".format(area_label))
+
+```
+xarrays plot function (built on matplotlib) can then be used to visualise the layers (classified surface above, albedo map below).
+
+```
+dataset.classified.plot()
+dataset.albedo.plot()
+```
+![xarray's plot function output: top = classified surface, bottom = albedo map](/assets/images/Sentinel_map.jpg)
```
Comment (ptheywood) [inline _posts/2019-02-08-xarray-for-earth.md:128]: These images are currently rendering as incredibly large. I'll fix this with some CSS and a separate pull request so we can keep the markdown clean. (I.e. don't worry about this).

File: _posts/2019-02-08-xarray-for-earth.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
+---
+title: xarray for Earth scientists
+author: Joseph Cook
+slug: xarray-for-earth
+date: 2019-02-08 11:48:00 UTC+01:00
+---
+
+# xarray for Earth Science
```
Comment (ptheywood) [inline _posts/2019-02-08-xarray-for-earth.md]: With how we render blog posts the (almost) replicated title seems unnecesary? 

image

**Review by ptheywood — CHANGES_REQUESTED** [review]
See the comments for minor changes.

File: _posts/2019-02-08-xarray-for-earth.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+from sklearn.externals import joblib
+
+vals = np.zeros([9,5490,5490]) #set up array
+bands = ['02','03','04','05','05','07','08','11','12'] # Band numbers
+for i in np.arange(0,len(bands),1): #loop through bands in NetCDF file
+    S2BX = xr.open_dataset(str(img_path+'B'+bands[i]+'.nc')) # open only the relevant band from the NetCDF
+    S2BXarr = S2BX.to_array() # send the band to a numpy array 
+    S2BXvals = np.array(S2BXarr.variable.values[1]) # extract only the reflectance values
+    S2BXvals = S2BXvals.astype(float) #convert to float
+    S2vals[i,:,:] = S2BXvals # append to 'master' ND array
```
Comment (willfurnass) [inline _posts/2019-02-08-xarray-for-earth.md:55]: `S2vals` undefined at this point.

File: _posts/2019-02-08-xarray-for-earth.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+vals = vals/10000 # correct unit from S2 L2A data to reflectance between 0-1
```
Comment (willfurnass) [inline _posts/2019-02-08-xarray-for-earth.md]: `vals /= 10000` more efficient

File: _posts/2019-02-08-xarray-for-earth.md (hunk 5 above)
Comment (ptheywood) [reply _posts/2019-02-08-xarray-for-earth.md:128]: This is fixed as of #61

**Comment by jmcook1186** [comment]
Hi - thanks for the comments so far - all amended in the second commit

**Review by ptheywood — APPROVED** [review]

**Comment by willfurnass** [comment]
Thanks for this @jmcook1186 !

**Pull request merged**